### PR TITLE
Finish Reverse DNS lookup

### PR DESCRIPTION
**Pull request by thorncorona**
# Description

DNS lookup code. 

```
InetAddress.getByName(ip).getCanonicalHostName();
```

Java's native implementation is very slow.
- 15s-20s to fail resolution
- 30-100ms to successfully resolve.

Out of sites

```
[
  'google.com',
  'reddit.com',
  'nytimes.com',
  'reddit.com',
  'facebook.com',
  'python.org',
  'imgur.com',
  'wikipedia.org',
  'youtube.com',
  'stackoverflow.com',
]
```
about 40% simply returned back their IP. 

My solution: 

- Resolve IP to DNS authority/server using Google DoH DNS
- Detect whether DNS returned authority
- Return response

- 30-50ms on average to resolve
- 90% takes 100ms to resolve
- No fail resolution penalties

PR for #17 
# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation made
- [x] New and existing unit tests pass locally with changes

# Tests Performed:

- [x] Upload data to Datastore
- [x] Retrieve data from Datastore
- [x] Map Visualization 
- [x] Network Map Visualization 
- [x] List other manual test cases performed

### CODE REVIEW

**Review by gpromanov — APPROVED** [review]

File: src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+  }
+
+  public String dnsRequest(String ip) throws Exception {
+    List<String> reverseIP = Arrays.asList(ip.split("\\."));
+    Collections.reverse(reverseIP);
+    String reverseIPString = String.join(".", reverseIP);
+
+    StringBuilder sb = new StringBuilder();
+    URL yahoo = new URL(String.format("https://dns.google.com/resolve?name=%s.in-addr.arpa&type=PTR", reverseIPString));
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java]: why do you want to call this `yahoo`?

File: src/main/java/com/google/netpcapanalysis/models/ReverseDNS.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  public boolean server;
+  public String record;
+
+  public ReverseDNS() {
+    authority = false;
+    server = false;
+  }
+
+  public ReverseDNS(String record, boolean server, boolean authority) {
+
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/models/ReverseDNS.java:37]: did you forget to add the implementation? or you dont need this overloaded constructor?

File: src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+      GoogleDNS res = new Gson().fromJson(request, GoogleDNS.class);
+
+      String data;
+      ReverseDNS rdns = new ReverseDNS();
+      if (res.Answer != null) {
+        data = res.Answer[res.Answer.length - 1].data;
+        rdns.server = true;
+      } else if (res.Authority != null) {
+        data = res.Authority[res.Authority.length - 1].data;
+        rdns.authority = true;
```
Comment (arunk054) [inline src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java]: typically you would use `setters` so in a large code base you can find where all this value  gets set.

**Review by arunk054 — APPROVED** [review]
Looks good!

**Comment by arunk054** [comment]
You should be able to merge this to master then rebase your `map_viz` to master.

File: src/main/java/com/google/netpcapanalysis/models/ReverseDNS.java (hunk 2 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/models/ReverseDNS.java:37]: I went ahead and filled it out

File: src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java (hunk 1 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java]: Scripted it out, forgot to change. Will change it to something more sensible

File: src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java (hunk 3 above)
Comment (thorncorona) [reply src/main/java/com/google/netpcapanalysis/dao/ReverseDNSLookupDaoImpl.java]: Changed

**Comment by thorncorona** [comment]
Rebased.

**Pull request merged**